KAFKA-4829: Improve log4j on Streams thread / task-level

guozhangwang opened the pull request:
These are the following improvements I made:

1. On stream thread level, INFO will be demonstrating `Completed xx tasks in yy ms` or `Completed rebalance with xx state in yy ms`, 
2. On Stream thread cache level, INFO on `Flushed xx records`.
3. On Stream thread level, DEBUG on internal batched operations like `created xx tasks`, and TRACE on individual operation like `created x task`.
4. Also using `isTraceEnabled` on the critical path to reduce overhead of creating `Object[]`.
5. Minor cleanups in the code.

guozhangwang committed:
improve log4j levels

guozhangwang commented:
@enothereska @mjsax @bbejeck for reviews.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5364/
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5379/
Test PASSed (JDK 7 and Scala 2.11).

guozhangwang commented:
retest this please.

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -255,7 +254,7 @@ void retryWithBackoff(final Map<TaskId, Set<TopicPartition>> tasksToBeCreated, f
                         it.remove();
                         backoffTimeMs = 50L;
                         if (retryingTasks.remove(taskId) && log.isWarnEnabled()) {
```
Nit: Should this be `log.isTraceEnabled()` ?

guozhangwang replied:
Ack.

bbejeck commented:
one minor comment otherwise LGTM.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5383/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5368/
Test PASSed (JDK 8 and Scala 2.12).

kamalcph commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTask.java:
```diff
@@ -137,7 +137,7 @@ public void close(final boolean clean) {
     public List<ConsumerRecord<byte[], byte[]>> update(final TopicPartition partition,
                                                        final List<ConsumerRecord<byte[], byte[]>> records) {
-        log.debug("{} Updating standby replicas of its state store for partition [{}]", logPrefix, partition);
+        log.trace("{} Updating standby replicas of its state store for partition [{}]", logPrefix, partition);
```
@guozhangwang 

Why the `if (log.isTraceEnabled())` call is not used in all the places?

guozhangwang replied:
I only added it for critical paths. If we want to do that everywhere across the board (i.e. beyond Streams) I think it would require a larger and separate pr.

ijuma replied:
Hmm, it seems like the `log.isTraceEnabled()` checks are not useful in some of the cases at least. If you pass up to 2 parameters, there is no benefit. See the underlying code:

```java
        if (isTraceEnabled()) {
            FormattingTuple ft = MessageFormatter.format(format, arg1, arg2);
            logger.log(FQCN, traceCapable ? Level.TRACE : Level.DEBUG, ft.getMessage(), ft.getThrowable());
        }
```

For more than 2 parameters (it would be nice if slf4j would have overloads for more parameters), there is an array allocation, which is generally pretty cheap as well.

nfo commented on streams/src/main/java/org/apache/kafka/streams/state/internals/ThreadCache.java:
```diff
@@ -210,10 +212,12 @@ private void maybeEvict(final String namespace) {
+            numEvicted++;
         }
+
+        log.info("{} Evicted {} entries from cache {}", logPrefix, numEvicted, namespace);
```
This line addresses my concerns. Thanks @guozhangwang.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -184,11 +184,14 @@ public boolean process() {
             final TopicPartition partition = recordInfo.partition();
 
-            log.trace("{} Start processing one record [{}]", logPrefix, record);
+            if (log.isTraceEnabled())
```
Nit: use `{ }` -- also for all other `if`s

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -710,7 +704,13 @@ public void apply(final StreamTask task) {
+
                     commitOne(task);
+
+                    log.info("{} Committed active task {} in {}ms",
```
`INFO` seems rather chatty to me. `DEBUG` or even `TRACE` might be better.

guozhangwang replied:
This is actually a top-level event, that user requested commit (note we have two types of such events: either commit interval has elapsed in which case all tasks are committed; or user called `commit` in which case only this single task is committed). We are using `INFO` for all such top-level events.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -771,11 +771,14 @@ private long adjustRecordsProcessedBeforeCommit(final long prevRecordsProcessedB
 
             commitAll();
+
+            log.info("{} Committed all active tasks {} and standby tasks {} in {}ms",
```
As above.

guozhangwang replied:
I think this is better stayed as `INFO`, and actually we are even more `chatty` that we have one entry per task, whereas now we are using one line indicating how much time has been elapsed for committing all tasks. If we want to make it DEBUG then I agree both this and the above should be debugging as you suggested, but my personal preference is keep it `INFO` since such events are quite important.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -846,6 +848,8 @@ private void maybeUpdateStandbyTasks() {
 
                     standbyRecords = remainingStandbyRecords;
+
+                    log.info("{} Updated standby tasks {} in {}ms", logPrefix, standbyTasks.keySet(), time.milliseconds() - now);
```
As above.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -1342,7 +1345,7 @@ private void addStandbyTasks(final long start) {
         final Map<TaskId, Set<TopicPartition>> newStandbyTasks = new HashMap<>();
 
-        log.info("{} Adding assigned standby tasks {}", logPrefix, partitionAssignor.activeTasks());
+        log.debug("{} Adding assigned standby tasks {}", logPrefix, partitionAssignor.activeTasks());
```
`partitionAssignor.activeTasks()` -> `partitionAssignor.standbyTasks()`

guozhangwang replied:
Good catch!

mjsax replied:
Was reported at the mailing list. :) There is actually a PR for it in the mean time (was not there when I put the comment): #3362

mjsax reviewed (commented):
Some minor comments.

It still feels a littler random when we use which logging level. IMHO, it would be good to define some rules  when to use which level. But that's beyond this PR.

guozhangwang committed:
github comments

guozhangwang commented:
Addressed comments. Per @mjsax suggestion I have downgraded thread-level global events to debug and corresponding debug level entries to trace levels.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5461/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5446/
Test PASSed (JDK 8 and Scala 2.12).

guozhangwang committed:
Some more tweaks on the log4j

guozhangwang commented:
I tweaked the log4j entries a bit more after running some example code while watching the different levels of log4j outputs closely. Specifically I decided to make the periodic committing as well as per-user-requested committing information still at INFO level (cc @mjsax ), while changing some internal partition assignor information from INFO to DEBUG.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5465/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5450/
Test PASSed (JDK 8 and Scala 2.12).

enothereska commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java:
```diff
@@ -153,7 +153,7 @@ public void restore() {
         } finally {
             consumer.assign(Collections.<TopicPartition>emptyList());
-            log.debug("{} Took {} ms to restore all active states", logPrefix, time.milliseconds() - start);
+            log.info("{} Completed restore all active states from changelog topics {} in {}ms ", logPrefix, needsRestoring.keySet(), time.milliseconds() - start);
```
Not sure if `info` is needed. With 100 state stores will get quite a bit of messages. isn't this just for debugging?

guozhangwang replied:
This is not one entry per state store, but one entry for all stores.

In this patch I tried to make all INFO level logging to be on StreamThread level and only for aggregated events.

mjsax reviewed (approved):
LGTM

stevenschlansker reviewed (approved):
Thank you for improving this!  Looking forward to the next release

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -184,11 +184,16 @@ public boolean process() {
             final TopicPartition partition = recordInfo.partition();
 
-            log.trace("{} Start processing one record [{}]", logPrefix, record);
+            if (log.isTraceEnabled()) {
```
You can remove the `if(log.isTraceEnabled()){..}` here and below as this already taken care of via the library. So there is no benefit in adding it here

guozhangwang replied:
Makes sense,

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -1245,7 +1243,7 @@ protected StreamTask createStreamTask(final TaskId id, final Collection<TopicPar
                 createProducer(id));
         } finally {
-            log.info("{} Created active task {} with assigned partitions {}", logPrefix, id, partitions);
+            log.trace("{} Created active task {} with assigned partitions {}", logPrefix, id, partitions);
```
should this really be trace level? I'd have though debug would be ok?

guozhangwang replied:
Note this is one log entry per task. On its caller (line 1341) we do have a debug-level entry:

```
 log.debug("{} Adding assigned standby tasks {}", logPrefix, partitionAssignor.standbyTasks());
```

dguy reviewed (commented):
Thanks @guozhangwang i left a couple of minor comments

guozhangwang committed:
rebase from trunk

guozhangwang committed:
Damian's comments

guozhangwang commented:
@dguy Thanks for the reviews, updated accordingly.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5586/
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5600/
Test PASSed (JDK 7 and Scala 2.11).

guozhangwang committed:
rebase from trunk and downgrade commit to debug

guozhangwang commented:
@dguy could you give it a final spin?

dguy reviewed (approved):
LGTM

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5662/
Test PASSed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5676/
Test PASSed (JDK 7 and Scala 2.11).

guozhangwang commented:
Merged to trunk, and cherry-picked with conflicts resolved to 0.11.0.
